src/main/java/com/microsoft/graph/authentication/TokenCredentialAuthProvider.java

```diff
@@ -0,0 +1,141 @@
+        this(scopes, Duration.ofMinutes(DEFAULT_TOKEN_TIMEOUT), tokenCredential);
+    }
+
+    /**
+     * Creates an Authentication provider using a TokenCredential and list of scopes
+     *
+     * @param tokenCredential Credential object inheriting the TokenCredential interface used to instantiate the Auth Provider
+     * @param scopes Specified desired scopes of the Auth Provider
+     * @param tokenObtentionTimeout Maximum time to wait for token obtention. Default 10 minutes. Use lower value on application with stable connectivity and no user interactions.
+     */
+    public TokenCredentialAuthProvider(@Nonnull final List<String> scopes, @Nonnull final Duration tokenObtentionTimeout, @Nonnull final TokenCredential tokenCredential) {
+        if(tokenCredential == null) {
+            throw new IllegalArgumentException("tokenCredential parameter cannot be null.");
+        }
+        if(tokenObtentionTimeout == null) {
+            throw new IllegalArgumentException("tokenObtentionTimout parameter cannot be null");
+        }
+        if(scopes == null || scopes.isEmpty()) {
+            throw new IllegalArgumentException("scopes parameter cannot be null or empty");
+        }
+        this.context = new TokenRequestContext();
+        this.context.setScopes(scopes);
+        this.tokenCredential = tokenCredential;
+        this.tokenBlockTimeout = tokenObtentionTimeout;
+    }
+
+    /**
+     * Authenticates the request
+     *
+     * @param request the request to authenticate
+     */
+    @Override
+    public void authenticateRequest(@Nonnull final IHttpRequest request) throws AuthenticationException {
+        if(request == null) {
+            throw new IllegalArgumentException("request parameter cannot be null.");
+        }
+        final URL requestUrl = request.getRequestUrl();
+        if(requestUrl != null && ShouldAuthenticateRequest(requestUrl.toString())) {
+            request.addHeader(InternalAuthConstants.AUTHORIZATION_HEADER, InternalAuthConstants.BEARER + getAccessToken());
+        }
+    }
+
+    /**
+     * Authenticates the request
+     *
+     * @param request the request to authenticate
+     * @return Request with Authorization header added to it
+     */
+    @Override
+    @Nonnull
+    public Request authenticateRequest(@Nonnull final Request request) throws AuthenticationException {
+        if(request == null) {
+            throw new IllegalArgumentException("request parameter cannot be null.");
+        }
+        final HttpUrl requestUrl = request.url();
+        if(requestUrl != null && ShouldAuthenticateRequest(requestUrl.toString())) {
+            return request.newBuilder()
+                    .addHeader(InternalAuthConstants.AUTHORIZATION_HEADER, InternalAuthConstants.BEARER + getAccessToken())
+                    .build();
+        } else {
+            return request;
+        }
+    }
+
```

Comment:
I suggest that the logic to verify the url Authentication can be moved to another file. It could be reused and not specific to the TokenCredential class.

Reply:
great minds connect! I thought about something along these lines yesterday after signing off. I just pushed the latest change that include this aspect as well as a some more refactoring. Overall it improves usability, separation of concerns and testability. Thanks for the suggestion a let me know what you think of this new iteration!